File: py/board.py
import random


class Board:
    def __init__(self, width, height, gen, tie_breaker=0):
        self.width = width
        self.height = height
        self.gen = gen
        self.board = []
        self.effectiveness = gen.get_effectiveness()
        self.types = gen.get_types()
        self.histogram = [0 for _ in range(len(gen.get_types()))]
        self.tie_breaker = tie_breaker

        for i in range(height):
            self.board.append([])
            for j in range(width):
                type_ = random.randint(0,len(gen.get_types())-1)
                self.histogram[type_] = self.histogram[type_] + 1
                self.board[i].append(type_)

    def clone(self):
        new_board = Board(self.width, self.height, self.gen)
        for i in range(self.width):
            for j in range(self.height):
                new_board.board[i][j] = self.board[i][j]
        return new_board

    def print(self):
        for i in range(self.height):
            for j in range(self.width):
                print(self.types[self.board[i][j]], end=', ')
            print()
    
    def get_next_state(self):
        new_board = self.clone()
        for x in range(self.width):
            for y in range(self.height):
                defn = self.board[x][y]
                looseCount = 0
                winners = []
                for i in range(-1,2):
                    for j in range(-1,2):
                        if (i == 0 and j == 0) or (x+i < 0 or x+i >= self.width or y+j < 0 or y+j >= self.height):
                            continue
                        att = self.board[x+i][y+j]
                        if self.effectiveness[att][defn] == 2 or (self.effectiveness[att][defn] == 1 and random.random() <= self.tie_breaker):
                            looseCount = looseCount + 1
                            winners.append(att)
                if (looseCount > 0):
                    winner = random.choice(winners)
                    new_board.board[x][y] = winner
                    self.histogram[winner] += 1
                    self.histogram[defn] -= 1
        self.board = new_board.board

File: csharp/lib/Board.cs
using ConwaysPokemon.Gens;
namespace ConwaysPokemon;

public class Board
{
    public int Width { get; set; }
    public int Height { get; set; }
    public Int32[,] Cells { get; set; }
    public IGeneration Generation { get; set; }

    public Board(int width, int height, IGeneration gen)
    {
        Random r = new Random();
        Width = width;
        Height = height;
        Cells = new Int32[width, height];
        for (int i = 0; i < width; i++)
        {
            for (int j = 0; j < height; j++)
            {
                Cells[i, j] = r.Next(0, gen.Types.Length);
            }
        }
        Generation = gen;
    }
    public void NextState()
    {
        Random r = new Random();
        for (int x = 0; x < Width; x++)
        {
            for (int y = 0; y < Height; y++)
            {
                int defender = Cells[x, y];
                List<int> winners = new List<int>();
                for (int i = -1; i < 2; i++)
                {
                    for (int j = -1; j < 2; j++)
                    {
                        if ((i == 0 && j == 0) || (x + i < 0 || x + i >= Width || y + j < 0 || y + j >= Height))
                        {
                            continue;
                        }
                        else
                        {
                            int attacker = Cells[x + i, y + j];
                            if (Generation.Effectiveness[attacker][defender] == 2)
                            {
                                winners.Add(attacker);
                            }
                        }
                    }
                }
                if (winners.Count > 0)
                {
                    Cells[x, y] = winners[r.Next(0, winners.Count)];
                }
            }
        }
    }
}